File: Union-Find/src/percolation.cpp
#include "include/percolation.h"
#include <iostream>

Percolation::Percolation(int n): n(n), n_sites(n*n), wqupc(n*n+2){
    open_sites.assign(n_sites+1, false);
}

void Percolation::validate(int row, int col){
    if(row > n || col > n || row < 0 || col < 0){
       std::cerr << "\nInvalid element index. Values for rows and cols must be between " <<
			"1 and " << n << '\n' << std::endl;
       std::cout << row << " " << col << std::endl;
		exit(EXIT_FAILURE); 
    }
}

int Percolation::row_col_to_site(int row, int col){
    return n*(row - 1) + col;
}

bool Percolation::is_open(int row, int col){
    validate(row, col);
    return open_sites[row_col_to_site(row, col)];
}

bool Percolation::is_open(int site){
    return open_sites[site];
}

int Percolation::number_of_open_sites(){
    return n_open_sites;
}

bool Percolation::percolates(){
    return wqupc.connected(0, n_sites+1);
}

void Percolation::open(int row, int col){
    validate(row, col);
    if(!is_open(row, col)){
        link_adjacent_sites(row, col);
        open_sites[row_col_to_site(row, col)] = true;
        n_open_sites++;
    }
}

void Percolation::link_adjacent_sites(int row, int col){
    validate(row, col);

    int cur_site = row_col_to_site(row, col);

    if(row != 1 && row !=n && col != n && col != 1){

        int top_site = row_col_to_site(row-1, col);
        int bottom_site = row_col_to_site(row+1, col);
        int left_site = row_col_to_site(row, col-1);
        int right_site = row_col_to_site(row, col+1);

        if(is_open(top_site)){
            wqupc.join(top_site, cur_site); 
        }
        if(is_open(bottom_site)){
            wqupc.join(bottom_site, cur_site); 
        }
        if(is_open(left_site)){
            wqupc.join(left_site, cur_site); 
        }
        if(is_open(right_site)){
            wqupc.join(right_site, cur_site); 
        }

    }

    else if(row == 1){

        wqupc.join(cur_site, 0);

        int bottom_site = row_col_to_site(row+1, col);
        int right_site = row_col_to_site(row, col+1);
        int left_site = row_col_to_site(row, col-1);

        if(is_open(bottom_site)){
            wqupc.join(cur_site, bottom_site);
        }
        if(col != n){
            if(is_open(right_site)){
                wqupc.join(cur_site, right_site);
            }
        }
        if(col != 1){
            if(is_open(left_site)){
                wqupc.join(cur_site, left_site);
            }
        }
    }

    else if(row == n){

        wqupc.join(cur_site, n_sites + 1);

        int top_site = row_col_to_site(row-1, col);
        int right_site = row_col_to_site(row, col+1);
        int left_site = row_col_to_site(row, col-1);

        if(is_open(top_site)){
            wqupc.join(cur_site, top_site);
        }
        if(col != n){
            if(is_open(right_site)){
                wqupc.join(cur_site, right_site);
            }
        }
        if(col != 1){
            if(is_open(left_site)){
                wqupc.join(cur_site, left_site);
            }
        }
    }

    else{

        int top_site = row_col_to_site(row-1, col);
        int bottom_site = row_col_to_site(row+1, col);
        int right_site = row_col_to_site(row, col+1);
        int left_site = row_col_to_site(row, col-1);

        if(is_open(top_site)){
            wqupc.join(cur_site, top_site);
        }
        if(is_open(bottom_site)){
            wqupc.join(cur_site, bottom_site);
        }
        if(col == 1){
            if(is_open(right_site)){
                wqupc.join(cur_site, right_site);
            }
        }
        if(col == n){
            if(is_open(left_site)){
                wqupc.join(cur_site, left_site);
            }
        }
    }
}

File: UnionFind/percolation.py
from wquf import WeightQuickUF


class Percolation:
    def __init__(self, n):
        self.size = n * n
        self.n = n
        self.n_open_sites = 0

        self.open_sites = [0 for i in range(self.size + 2)]
        self.open_sites[0] = 1
        self.open_sites[self.size + 1] = 1

        self.wquf = WeightQuickUF(self.size + 2)

    def rowColToSite(self, row, col):
        return 1 + (row - 1) * self.n + (col - 1)

    def isPercolates(self):
        return self.wquf.isConnected(0, self.size + 1)

    def isOpen(self, row, col):
        site = self.rowColToSite(row, col)
        return self.open_sites[site]

    def connectAdjacentSites(self, row, col):
        site = self.rowColToSite(row, col)

        if (row == self.n):

            self.wquf.connect(site, self.size + 1)

            if (col != 1):
                if (self.isOpen(row, col - 1)):
                    self.wquf.connect(site, self.rowColToSite(row, col - 1))

            if (col != self.n):
                if (self.isOpen(row, col + 1)):
                    self.wquf.connect(site, self.rowColToSite(row, col + 1))

            if (self.isOpen(row - 1, col)):
                self.wquf.connect(site, self.rowColToSite(row - 1, col))

        elif (row == 1):

            self.wquf.connect(site, 0)

            if (col != 1):
                if (self.isOpen(row, col - 1)):
                    self.wquf.connect(site, self.rowColToSite(row, col - 1))

            if (col != self.n):
                if (self.isOpen(row, col + 1)):
                    self.wquf.connect(site, self.rowColToSite(row, col + 1))

            if (self.isOpen(row + 1, col)):
                self.wquf.connect(site, self.rowColToSite(row + 1, col))

        elif (row != 1 and row != self.n and col != 1 and col != self.n):

            if (self.isOpen(row - 1, col)):
                self.wquf.connect(site, self.rowColToSite(row - 1, col))

            if (self.isOpen(row + 1, col)):
                self.wquf.connect(site, self.rowColToSite(row + 1, col))

            if (self.isOpen(row, col - 1)):
                self.wquf.connect(site, self.rowColToSite(row, col - 1))

            if (self.isOpen(row, col + 1)):
                self.wquf.connect(site, self.rowColToSite(row, col + 1))

        else:
            if (col == 1):

                if (self.isOpen(row - 1, col)):
                    self.wquf.connect(site, self.rowColToSite(row - 1, col))

                if (self.isOpen(row + 1, col)):
                    self.wquf.connect(site, self.rowColToSite(row + 1, col))

                if (self.isOpen(row, col + 1)):
                    self.wquf.connect(site, self.rowColToSite(row, col + 1))

            if (col == self.n):

                if (self.isOpen(row - 1, col)):
                    self.wquf.connect(site, self.rowColToSite(row - 1, col))

                if (self.isOpen(row + 1, col)):
                    self.wquf.connect(site, self.rowColToSite(row + 1, col))

                if (self.isOpen(row, col - 1)):
                    self.wquf.connect(site, self.rowColToSite(row, col - 1))

    def open(self, row, col):
        site = self.rowColToSite(row, col)
        if (not self.isOpen(row, col)):
            self.open_sites[site] = 1
            self.n_open_sites += 1
            self.connectAdjacentSites(row, col)

    def print_array(self):
        self.wquf.print_array()
        i = 1
        while (i != self.size + 1):
            print(self.open_sites[i], end=" ")
            if (i % (self.n) == 0):
                print()
            i += 1
        print()
